File: app/src/main/java/dev/ragnarok/fenrir/push/message/NewPostPushMessage.java
package dev.ragnarok.fenrir.push.message;

import static java.lang.Integer.parseInt;
import static dev.ragnarok.fenrir.push.NotificationUtils.configOtherPushNotification;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;

import androidx.core.app.NotificationCompat;

import com.google.firebase.messaging.RemoteMessage;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.MainActivity;
import dev.ragnarok.fenrir.longpoll.AppNotificationChannels;
import dev.ragnarok.fenrir.longpoll.NotificationHelper;
import dev.ragnarok.fenrir.push.NotificationScheduler;
import dev.ragnarok.fenrir.push.OwnerInfo;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.settings.theme.ThemesController;
import dev.ragnarok.fenrir.util.Logger;
import dev.ragnarok.fenrir.util.Utils;

public class NewPostPushMessage {
    private static final Pattern PATTERN_WALL_POST = Pattern.compile("vk.com/(?:[\\w.\\d]+\\?(?:[\\w=&]+)?w=)?wall(-?\\d*)_(\\d*)");

    private final int accountId;
    private final String url;
    private final String body;
    private final String title;

    public NewPostPushMessage(int accountId, RemoteMessage remote) {
        this.accountId = accountId;
        Map<String, String> data = remote.getData();
        url = data.get("url");
        body = data.get("body");
        title = data.get("title");
    }

    @SuppressLint("CheckResult")
    public void notifyIfNeed(Context context) {
        if (url == null) {
            Logger.wtf("NewPostPushMessage", "url is NULL!!!");
            return;
        }

        if (!Settings.get()
                .notifications()
                .isNewPostsNotificationEnabled()) {
            return;
        }

        Matcher matcher = PATTERN_WALL_POST.matcher(url);
        if (matcher.find()) {
            Context app = context.getApplicationContext();
            OwnerInfo.getRx(app, accountId, parseInt(matcher.group(1)))
                    .subscribeOn(NotificationScheduler.INSTANCE)
                    .subscribe(ownerInfo -> notifyImpl(app, ownerInfo.getAvatar()), throwable -> notifyImpl(app, null));
        } else {
            notifyImpl(context, null);
        }
    }

    private void notifyImpl(Context context, Bitmap bitmap) {
        NotificationManager nManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (Utils.hasOreo()) {
            nManager.createNotificationChannel(AppNotificationChannels.getNewPostChannel(context));
        }

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, AppNotificationChannels.getNewPostChannelId())
                .setSmallIcon(R.drawable.client_round)
                .setContentTitle(title)
                .setContentText(body)
                .setColor(ThemesController.INSTANCE.toastColor(false))
                .setStyle(new NotificationCompat.BigTextStyle().bigText(body))
                .setAutoCancel(true);
        if (bitmap != null) {
            builder.setLargeIcon(bitmap);
        }

        builder.setPriority(NotificationCompat.PRIORITY_HIGH);

        Intent intent = new Intent(context, MainActivity.class);
        intent.setAction(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(url));
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        PendingIntent contentIntent = PendingIntent.getActivity(context, url.hashCode(), intent, Utils.makeMutablePendingIntent(PendingIntent.FLAG_CANCEL_CURRENT));
        builder.setContentIntent(contentIntent);
        Notification notification = builder.build();

        configOtherPushNotification(notification);

        nManager.notify(url, NotificationHelper.NOTIFICATION_NEW_POSTS_ID, notification);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/push/message/NewPostPushMessage.kt
package dev.ragnarok.fenrir.push.message

import android.annotation.SuppressLint
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import androidx.core.app.NotificationCompat
import com.google.firebase.messaging.RemoteMessage
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.MainActivity
import dev.ragnarok.fenrir.longpoll.AppNotificationChannels.getNewPostChannel
import dev.ragnarok.fenrir.longpoll.AppNotificationChannels.newPostChannelId
import dev.ragnarok.fenrir.longpoll.NotificationHelper
import dev.ragnarok.fenrir.push.NotificationScheduler.INSTANCE
import dev.ragnarok.fenrir.push.NotificationUtils.configOtherPushNotification
import dev.ragnarok.fenrir.push.OwnerInfo
import dev.ragnarok.fenrir.push.OwnerInfo.Companion.getRx
import dev.ragnarok.fenrir.settings.Settings.get
import dev.ragnarok.fenrir.settings.theme.ThemesController.toastColor
import dev.ragnarok.fenrir.util.AppPerms
import dev.ragnarok.fenrir.util.Logger.wtf
import dev.ragnarok.fenrir.util.Utils.hasOreo
import dev.ragnarok.fenrir.util.Utils.makeMutablePendingIntent
import java.util.regex.Pattern

class NewPostPushMessage {
    private var accountId = 0
    private var url: String? = null
    private var body: String? = null
    private var title: String? = null

    @SuppressLint("CheckResult")
    fun notifyIfNeed(context: Context) {
        val pUrl = url
        if (pUrl == null) {
            wtf("NewPostPushMessage", "url is NULL!!!")
            return
        }
        if (!get()
                .notifications()
                .isNewPostsNotificationEnabled
        ) {
            return
        }
        val matcher = PATTERN_WALL_POST.matcher(pUrl)
        if (matcher.find()) {
            matcher.group(1)?.let {
                val app = context.applicationContext
                getRx(app, accountId, it.toInt())
                    .subscribeOn(INSTANCE)
                    .subscribe({ ownerInfo: OwnerInfo ->
                        notifyImpl(
                            app,
                            ownerInfo.avatar
                        )
                    }) { notifyImpl(app, null) }
            } ?: notifyImpl(context, null)
        } else {
            notifyImpl(context, null)
        }
    }

    private fun notifyImpl(context: Context, bitmap: Bitmap?) {
        val nManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        if (hasOreo()) {
            nManager.createNotificationChannel(getNewPostChannel(context))
        }
        val builder = NotificationCompat.Builder(context, newPostChannelId)
            .setSmallIcon(R.drawable.client_round)
            .setContentTitle(title)
            .setContentText(body)
            .setColor(toastColor(false))
            .setStyle(NotificationCompat.BigTextStyle().bigText(body))
            .setAutoCancel(true)
        if (bitmap != null) {
            builder.setLargeIcon(bitmap)
        }
        builder.priority = NotificationCompat.PRIORITY_HIGH
        val intent = Intent(context, MainActivity::class.java)
        intent.action = Intent.ACTION_VIEW
        intent.data = Uri.parse(url)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        val contentIntent = PendingIntent.getActivity(
            context,
            url.hashCode(),
            intent,
            makeMutablePendingIntent(PendingIntent.FLAG_CANCEL_CURRENT)
        )
        builder.setContentIntent(contentIntent)
        val notification = builder.build()
        configOtherPushNotification(notification)
        if (AppPerms.hasNotificationPermissionSimple(context)) {
            nManager.notify(url, NotificationHelper.NOTIFICATION_NEW_POSTS_ID, notification)
        }
    }

    companion object {
        private val PATTERN_WALL_POST =
            Pattern.compile("vk.com/(?:[\\w.\\d]+\\?(?:[\\w=&]+)?w=)?wall(-?\\d*)_(\\d*)")

        fun fromRemoteMessage(accountId: Int, remote: RemoteMessage): NewPostPushMessage {
            val message = NewPostPushMessage()
            message.accountId = accountId
            val data = remote.data
            message.url = data["url"]
            message.body = data["body"]
            message.title = data["title"]
            return message
        }
    }
}